src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2792,212 +3027,268 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
-         * the output arrays and clear those vectors
-         */
-        output_offset       = (size_t)(k)*max_iovec_len;
-        _mem_buf_offset     = mem_buf_offset + output_offset;
-        _target_file_offset = target_file_offset + output_offset;
-        _io_block_len       = io_block_len + output_offset;
-
-        memset(_mem_buf_offset, 0, (max_iovec_len * sizeof(*_mem_buf_offset)));
-        memset(_target_file_offset, 0, (max_iovec_len * sizeof(*_target_file_offset)));
-        memset(_io_block_len, 0, (max_iovec_len * sizeof(*_io_block_len)));
-
-        if (total_bytes == data_size) {
-            *n_subfiles_used = i;
-            goto done;
-        }
+        iovec_depth      = cur_max_iovec_depth;
+        num_full_stripes = iovec_depth;
 
-        if (total_bytes < data_size) {
-            int64_t num_full_stripes = iovec_depth;
+        if (subfile_idx == first_subfile_idx) {
+            is_first = true;
 
-            if (k == first_subfile) {
-                is_first = true;
+            /*
+             * Add partial segment length if not
+             * starting on a stripe boundary
+             */
+            if (start_length < stripe_size) {
+                subfile_bytes += start_length;
+                num_full_stripes--;
+            }
+        }
 
-                /*
-                 * Add partial segment length if not
-                 * starting on a stripe boundary
-                 */
-                if (start_length < stripe_size) {
-                    subfile_bytes += start_length;
+        if (subfile_idx == last_subfile_idx) {
+            is_last = true;
+
+            /*
+             * Add partial segment length if not
+             * ending on a stripe boundary
+             */
+            if (final_length < stripe_size) {
+                subfile_bytes += final_length;
+                if (num_full_stripes)
                     num_full_stripes--;
-                }
             }
+        }
 
-            if (k == last_subfile) {
-                is_last = true;
+        /* Account for subfiles with uniform segments */
+        if (!is_first && !is_last) {
+            bool thin_uniform_section = false;
 
+            if (last_subfile_idx >= first_subfile_idx) {
                 /*
-                 * Add partial segment length if not
-                 * ending on a stripe boundary
+                 * When a subfile has an index value that is greater
+                 * than both the starting subfile and ending subfile
+                 * indices, it is a "thinner" section with a smaller
+                 * I/O vector depth.
                  */
-                if (final_length < stripe_size) {
-                    subfile_bytes += final_length;
-                    if (num_full_stripes)
-                        num_full_stripes--;
-                }
+                thin_uniform_section = (subfile_idx > first_subfile_idx) && (subfile_idx > last_subfile_idx);
```

Comment:
Wouldn't this also be the case if it was below both? Also don't you only need to check if it's greater than last_subfile_idx? (And if it's less than first_subfile_idx in the case I mentioned)

Reply:
> Wouldn't this also be the case if it was below both?

After thinking about this for a while and drawing some diagrams, I believe this is the case, yes, though it should just populate the I/O vectors with data that isn't used. Let me write a test for this case and verify, then I'll fix this.

> Also don't you only need to check if it's greater than last_subfile_idx?

Ah, yes that follows from the logic of `if (last_subfile_idx >= first_subfile_idx)`. Good catch.